server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/RegexMappingRoutesTest.java

```diff
@@ -0,0 +1,219 @@
+            .createWebAdminServer(new RegexMappingRoutes(memoryRecipientRewriteTable))
+            .start();
+
+        RestAssured.requestSpecification = WebAdminUtils.buildRequestSpecification(webAdminServer)
+            .setBasePath(RegexMappingRoutes.BASE_PATH)
+            .log(LogDetail.METHOD)
+            .build();
+    }
+
+    @AfterEach
+    void stop() {
+        webAdminServer.destroy();
+    }
+
+    @Test
+    void addRegexMappingShouldReturnNoContentWhenSuccess() {
+        with()
+            .body(
+                "{" +
+                "  \"source\": \"james@domain.tld\"," +
+                "  \"regex\": \"^[aeiou]\"" +
+                "}")
+            .post()
+        .then()
+            .statusCode(HttpStatus.NO_CONTENT_204)
+            .contentType(ContentType.JSON);
+
+        assertThat(memoryRecipientRewriteTable
+            .getStoredMappings(MappingSource.fromUser(User.fromUsername("james@domain.tld"))))
+            .containsOnly(Mapping.regex("^[aeiou]"));
+    }
+
+    @Test
+    void addRegexMappingShouldReturnBadRequestWhenBodyIsInvalid() {
+        with()
+            .body("Invalid body")
+            .post()
+        .then()
+            .statusCode(HttpStatus.BAD_REQUEST_400)
+            .contentType(ContentType.JSON);
+    }
+
+    @Test
+    void addRegexMappingShouldReturnBadRequestWhenSourceIsEmpty() {
+        with()
+            .body(
+                "{" +
+                "  \"source\": \"\"," +
+                "  \"regex\": \"^[aeiou]\"" +
+                "}")
+            .post()
+        .then()
+            .statusCode(HttpStatus.BAD_REQUEST_400)
+            .contentType(ContentType.JSON)
+            .body("message", is("Invalid `source` field."));
+    }
+
+    @Test
+    void addRegexMappingShouldAllowEmptyRegex() {
+        with()
+            .body(
+                "{" +
+                "  \"source\":\"james@domain.tld\"," +
+                "  \"regex\": \"\"" +
+                "}")
+            .post()
+        .then()
+            .statusCode(HttpStatus.NO_CONTENT_204)
+            .contentType(ContentType.JSON);
+    }
+
+    @Test
+    void addRegexMappingShouldReturnBadRequestWhenSourceAndRegexEmpty() {
+        with()
+            .body(
+                "{" +
+                "  \"source\": \"\"," +
+                "  \"regex\": \"\"" +
+                "}")
+            .post()
+        .then()
+            .statusCode(HttpStatus.BAD_REQUEST_400)
+            .contentType(ContentType.JSON)
+            .body("message", is("Invalid `source` field."));
+    }
+
+    @Test
+    void addRegexMappingShouldReturnBadRequestWhenSourceIsNull() {
+        with()
+            .body(
+                "{" +
+                "  \"source\": null," +
+                "  \"regex\": \"yourRegexExpression\"" +
+                "}")
+            .post()
+        .then()
+            .statusCode(HttpStatus.BAD_REQUEST_400)
+            .contentType(ContentType.JSON);
```

Comment:
What is the JSON returned by this exception? (and all 400 exceptions below?)

Reply:
I have clarified JSON content in all 400 exceptions.

Reply:
Talking about the json body here, not the content type :)